### PR TITLE
(initial) ML-DSA support

### PR DESCRIPTION
**Pull request by bigbrett**
Adds initial support for ML-DSA to wolfHSM. Note that this requires https://github.com/wolfSSL/wolfssl/pull/8177 for wolfCrypt "blind deserialization" of ML-DSA DER keys. **CI tests will fail until this merges.** You can manually test against this branch if desired.

This PR does not include DMA support, which will be required (due to large key/signature sizes) for this to be reasonable to use on real hardware. I'm working on that part now, but wanted initial support to be merged in first.

This PR also fixes a few bugs in the transport mem and comm context layers, where improperly sized buffers could result in buffer overflows.

### CODE REVIEW

File: src/wh_client_crypto.c [hunk 1]
```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
     return ret;
 }
 #endif /* WOLFSSL_CMAC */
+
+#ifdef HAVE_DILITHIUM
```
Comment (billphipps) [inline src/wh_client_crypto.c:2076]: Yuck.  This should be spelled HAVE_MLDSA or HAVE_FIPS204.  No change.  Just complaining

File: src/wh_client_crypto.c [hunk 2]
```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
...
+    return WH_ERROR_OK;
+}
+
+int wh_Client_MlDsaGetKeyId(MlDsaKey* key, whKeyId* outId)
+{
+    if (key == NULL || outId == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+
+    *outId = (whKeyId)((intptr_t)key->devCtx);
```
Comment (billphipps) [inline src/wh_client_crypto.c:2094]:
```suggestion
    *outId = WH_DEVCTX_TO_KEYID(key->devCtx);
```

File: src/wh_client_crypto.c [hunk 3]
```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
...
+
+#ifdef HAVE_DILITHIUM
+
+int wh_Client_MlDsaSetKeyId(MlDsaKey* key, whKeyId keyId)
+{
+    if (key == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+
+    key->devCtx = (void*)((intptr_t)keyId);
```
Comment (billphipps) [inline src/wh_client_crypto.c:2084]:
```suggestion
    key->devCtx = WH_KEYID_TO_DEVCTX(keyId);
```

File: src/wh_client_crypto.c [hunk 4]
```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
...
+    }
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+    printf("[client] %s keyid:%x key:%p ret:%d label:%.*s\n", __func__, keyId,
+           key, ret, (int)label_len, label);
+#endif
+    return ret;
+}
+
+static int _MlDsaMakeKey(whClientContext* ctx, int size, int level,
```
Comment (billphipps) [inline src/wh_client_crypto.c:2167]: Can you add a declaration of this at the top?  I swear we were getting complaints about this in the HighTec compiler

File: src/wh_client_crypto.c [hunk 5]
```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
...
+    if (inout_key_id == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+
+    return _MlDsaMakeKey(ctx, size, level, inout_key_id, flags, label_len,
+                         label, NULL);
+}
+
+int wh_Client_MlDsaMakeExportKey(whClientContext* ctx, int level, MlDsaKey* key,
+                                 int size, WC_RNG* rng)
```
Comment (billphipps) [inline src/wh_client_crypto.c:2292]: RNG should be removed.  The server will use its own rng.

File: src/wh_client_crypto.c [hunk 6]
```diff
@@ -2072,4 +2072,492 @@ int wh_Client_CmacCancelableResponse(whClientContext* c, Cmac* cmac,
...
+        return WH_ERROR_BADARGS;
+    }
+
+    return _MlDsaMakeKey(ctx, size, level, NULL, WH_NVM_FLAGS_EPHEMERAL, 0,
+                         NULL, key);
+}
+
+
+int wh_Client_MlDsaSign(whClientContext* ctx, const byte* in, word32 in_len,
+                        byte* out, word32* inout_len, WC_RNG* rng,
```
Comment (billphipps) [inline src/wh_client_crypto.c:2304]: No rng here either.

File: src/wh_client_cryptocb.c [hunk 7]
```diff
@@ -727,6 +751,125 @@ static int _handleSha256Dma(wc_CryptoInfo* info, void* inCtx, whPacket* packet)
 #endif /* ! NO_SHA256 */
 
 
+#if defined(HAVE_FALCON) || defined(HAVE_DILITHIUM)
+static int _handlePqcSigKeyGen(whClientContext* ctx, wc_CryptoInfo* info)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+
+    /* Extract info parameters */
+    WC_RNG* rng  = info->pk.pqc_sig_kg.rng;
```
Comment (billphipps) [inline src/wh_client_cryptocb.c:760]: No need for client-side rng

File: src/wh_client_cryptocb.c [hunk 8]
```diff
@@ -727,6 +751,125 @@ static int _handleSha256Dma(wc_CryptoInfo* info, void* inCtx, whPacket* packet)
...
+static int _handlePqcSign(whClientContext* ctx, wc_CryptoInfo* info)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+
+    /* Extract info parameters */
+    const byte* in      = info->pk.pqc_sign.in;
+    word32      in_len  = info->pk.pqc_sign.inlen;
+    byte*       out     = info->pk.pqc_sign.out;
+    word32*     out_len = info->pk.pqc_sign.outlen;
+    WC_RNG*     rng     = info->pk.pqc_sign.rng;
```
Comment (billphipps) [inline src/wh_client_cryptocb.c:792]: No client-side rng needed

File: src/wh_comm.c [hunk 9]
```diff
@@ -108,6 +108,11 @@ int wh_CommClient_SendRequest(whCommClient* context, uint16_t magic,
         return WH_ERROR_BADARGS;
     }
 
+    /* Check if the data size is within allowed limits */
+    if (data_size > WOLFHSM_CFG_COMM_DATA_LEN) {
```
Comment (billphipps) [inline src/wh_comm.c:112]: How did we miss this??  I assume the lower-level does a sanity check but we should have caught it here.  Nice!

File: src/wh_server_crypto.c [hunk 10]
```diff
@@ -497,6 +513,70 @@ int wh_Server_CacheExportCurve25519Key(whServerContext* server, whKeyId keyId,
...
+int wh_Server_MlDsaKeyCacheImport(whServerContext* ctx, MlDsaKey* key,
+                                  whKeyId keyId, whNvmFlags flags,
+                                  uint16_t label_len, uint8_t* label)
+{
+    int            ret = WH_ERROR_OK;
+    uint8_t*       cacheBuf;
+    whNvmMetadata* cacheMeta;
+    uint16_t       der_size;
+
+    const uint16_t MAX_MLDSA_DER_SIZE = 5000;
```
Comment (billphipps) [inline src/wh_server_crypto.c:526]: Isn't this defined somewhere?

File: test/wolfhsm_cfg.h [hunk 11]
```diff
@@ -30,13 +30,13 @@
...
 
-#define WOLFHSM_CFG_COMM_DATA_LEN 1280
+#define WOLFHSM_CFG_COMM_DATA_LEN (1280 * 4)
 
 #define WOLFHSM_CFG_NVM_OBJECT_COUNT 30
 #define WOLFHSM_CFG_SERVER_KEYCACHE_COUNT 9
 #define WOLFHSM_CFG_SERVER_KEYCACHE_BUFSIZE 300
 #define WOLFHSM_CFG_SERVER_KEYCACHE_BIG_COUNT 2
-#define WOLFHSM_CFG_SERVER_KEYCACHE_BIG_BUFSIZE 1280
+#define WOLFHSM_CFG_SERVER_KEYCACHE_BIG_BUFSIZE WOLFHSM_CFG_COMM_DATA_LEN
```
Comment (billphipps) [inline test/wolfhsm_cfg.h:39]: This is a clever change.  :)

**Review by billphipps — APPROVED** [review]
Minor nits that can be tweaked moving forward.  Very tasty!

File: src/wh_client_crypto.c (hunk 2 above)
Comment (bigbrett) [reply src/wh_client_crypto.c:2094]: fixed in #86

File: src/wh_client_crypto.c (hunk 3 above)
Comment (bigbrett) [reply src/wh_client_crypto.c:2084]: fixed in #86

File: src/wh_client_crypto.c (hunk 4 above)
Comment (bigbrett) [reply src/wh_client_crypto.c:2167]: fixed in #86

File: src/wh_client_crypto.c (hunk 5 above)
Comment (bigbrett) [reply src/wh_client_crypto.c:2292]: fixed in #86

File: src/wh_client_crypto.c (hunk 6 above)
Comment (bigbrett) [reply src/wh_client_crypto.c:2304]: fixed in #86

File: src/wh_client_cryptocb.c (hunk 7 above)
Comment (bigbrett) [reply src/wh_client_cryptocb.c:760]: fixed in #86

File: src/wh_client_cryptocb.c (hunk 8 above)
Comment (bigbrett) [reply src/wh_client_cryptocb.c:792]: fixed in #86

File: src/wh_server_crypto.c (hunk 10 above)
Comment (bigbrett) [reply src/wh_server_crypto.c:526]: fixed in #86

**Pull request merged**